cmd/cosign/cli/trustedroot/trustedroot.go

```diff
@@ -0,0 +1,193 @@
+	CAIntermediates  string
+	CARoots          string
+	CertChain        string
+	Out              string
+	RekorURL         string
+	TSACertChainPath string
+}
+
+func (c *CreateCmd) Exec(ctx context.Context) error {
+	var fulcioCertAuthorities []root.CertificateAuthority
+	var timestampAuthorities []root.CertificateAuthority
+	rekorTransparencyLogs := make(map[string]*root.TransparencyLog)
+
+	if c.CertChain != "" {
+		fulcioAuthority, err := parsePEMFile(c.CertChain)
+		if err != nil {
+			return err
+		}
+		fulcioCertAuthorities = append(fulcioCertAuthorities, *fulcioAuthority)
+	} else if c.CARoots != "" {
+		roots, err := parseCerts(c.CARoots)
+		if err != nil {
+			return err
+		}
+
+		var intermediates []*x509.Certificate
+		if c.CAIntermediates != "" {
+			intermediates, err = parseCerts(c.CAIntermediates)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Here we're trying to "flatten" the x509.CertPool cosign was using
+		// into a trusted root with a clear mapping between roots and
+		// intermediates. Make a guess that if there are intermediates, there
+		// is one per root.
+
+		for i, rootCert := range roots {
+			var fulcioAuthority root.CertificateAuthority
+			fulcioAuthority.Root = rootCert
+			if i < len(intermediates) {
+				fulcioAuthority.Intermediates = []*x509.Certificate{intermediates[i]}
+			}
+			fulcioCertAuthorities = append(fulcioCertAuthorities, fulcioAuthority)
+		}
+	}
+
+	if c.RekorURL != "" {
+		rekorClient, err := rekor.NewClient(c.RekorURL)
+		if err != nil {
+			return fmt.Errorf("creating Rekor client: %w", err)
+		}
+
+		rekorPubKey, err := rekorClient.Pubkey.GetPublicKey(nil)
+		if err != nil {
+			return err
+		}
+
+		block, _ := pem.Decode([]byte(rekorPubKey.Payload))
+		if block == nil {
+			return errors.New("failed to decode public key of server")
+		}
+
+		pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+		if err != nil {
+			return err
+		}
+
+		keyHash := sha256.Sum256(block.Bytes)
+		keyID := base64.StdEncoding.EncodeToString(keyHash[:])
+
+		rekorTransparencyLog := root.TransparencyLog{
+			BaseURL:           c.RekorURL,
+			HashFunc:          crypto.SHA256,
+			ID:                keyHash[:],
+			PublicKey:         pub,
+			SignatureHashFunc: crypto.SHA256,
+		}
+
+		rekorTransparencyLogs[keyID] = &rekorTransparencyLog
+	}
+
+	if c.TSACertChainPath != "" {
+		timestampAuthority, err := parsePEMFile(c.TSACertChainPath)
+		if err != nil {
+			return err
+		}
+		timestampAuthorities = append(timestampAuthorities, *timestampAuthority)
+	}
+
+	newTrustedRoot, err := root.NewTrustedRoot(root.TrustedRootMediaType01,
+		fulcioCertAuthorities, nil, timestampAuthorities, rekorTransparencyLogs,
+	)
+	if err != nil {
+		return err
+	}
+
+	var trBytes []byte
+
+	trBytes, err = newTrustedRoot.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	if c.Out != "" {
+		err = os.WriteFile(c.Out, trBytes, 0600)
+		if err != nil {
+			return err
+		}
+	} else {
+		ui.Infof(ctx, string(trBytes))
```

Comment:
It looks like this writes to stderr rather than stdout, so it's not intuitive to capture it with a `>` redirect. Not super critical since you also have the `-out` option, just less intuitive to me.